mxcache-generation/src/test/java/com/maxifier/mxcache/template/impl/PInlineDependencyCache.template

```diff
@@ -68,6 +76,26 @@ public class #E#InlineDependencyCache extends #E#InlineCacheImpl implements Depe
             dependentNodes = new THashSet<Reference<DependencyNode>>();
         }
         dependentNodes.add(node.getSelfReference());
+        cleanupIfNeeded();
+    }
+
+    private void cleanupIfNeeded() {
+        if (dependentNodes.size() >= cleanupThreshold) {
+            for (Iterator<Reference<DependencyNode>> it = dependentNodes.iterator(); it.hasNext(); ) {
+                if (it.next().get() == null) {
+                    it.remove();
+                }
+            }
+            // It's important to increase cleanup threshold according to the number of elements in a set
+            // in order to maintain the balance between CPU-overhead and memory-overhead
+
+            // The cleanup has O(N) complexity, so doing this on addition of N new elements would lead to constant
+            // small overhead and thus would not affect the asymptotic behaviour of operations.
+
+            // The memory overhead could be significant but it's guaranteed that memory usage would not be more than
+            // 2 * peak memory usage for alive elements.
+            cleanupThreshold = dependentNodes.size() * 2;

```

Comment:
That might help, however:
- there are thousands of hundreds of dependency nodes, so this might hurt performance.
- it's hard to obtain the list of all dependency nodes.
- the memory footprint will still be high if allocation seed is high (e.g. if 1M of objects is allocated and then GC'e within the interval of cleanups).